modules/core/src/matrix_transform.cpp

```diff
@@ -516,92 +516,113 @@ flipHoriz( const uchar* src, size_t sstep, uchar* dst, size_t dstep, Size size,
+                for( ; i <= sz.width - 16; i += 16 )
+                {
+                    int t0 = ((int*)(srcTop + i))[0];
+                    int t1 = ((int*)(srcBottom + i))[0];
 
-                ((int*)(dst0 + i))[0] = t1;
-                ((int*)(dst1 + i))[0] = t0;
+                    ((int*)(dstTop + i))[0] = t1;
+                    ((int*)(dstBottom + i))[0] = t0;
 
-                t0 = ((int*)(src0 + i))[1];
-                t1 = ((int*)(src1 + i))[1];
+                    t0 = ((int*)(srcTop + i))[1];
+                    t1 = ((int*)(srcBottom + i))[1];
 
-                ((int*)(dst0 + i))[1] = t1;
-                ((int*)(dst1 + i))[1] = t0;
+                    ((int*)(dstTop + i))[1] = t1;
+                    ((int*)(dstBottom + i))[1] = t0;
 
-                t0 = ((int*)(src0 + i))[2];
-                t1 = ((int*)(src1 + i))[2];
+                    t0 = ((int*)(srcTop + i))[2];
+                    t1 = ((int*)(srcBottom + i))[2];
 
-                ((int*)(dst0 + i))[2] = t1;
-                ((int*)(dst1 + i))[2] = t0;
+                    ((int*)(dstTop + i))[2] = t1;
+                    ((int*)(dstBottom + i))[2] = t0;
 
-                t0 = ((int*)(src0 + i))[3];
-                t1 = ((int*)(src1 + i))[3];
+                    t0 = ((int*)(srcTop + i))[3];
+                    t1 = ((int*)(srcBottom + i))[3];
 
-                ((int*)(dst0 + i))[3] = t1;
-                ((int*)(dst1 + i))[3] = t0;
-            }
+                    ((int*)(dstTop + i))[3] = t1;
+                    ((int*)(dstBottom + i))[3] = t0;
+                }
 
-            for( ; i <= size.width - 4; i += 4 )
-            {
-                int t0 = ((int*)(src0 + i))[0];
-                int t1 = ((int*)(src1 + i))[0];
+                for( ; i <= sz.width - 4; i += 4 )
+                {
+                    int t0 = ((int*)(srcTop + i))[0];
+                    int t1 = ((int*)(srcBottom + i))[0];
 
-                ((int*)(dst0 + i))[0] = t1;
-                ((int*)(dst1 + i))[0] = t0;
+                    ((int*)(dstTop + i))[0] = t1;
+                    ((int*)(dstBottom + i))[0] = t0;
+                }
             }
-        }
 
-        for( ; i < size.width; i++ )
-        {
-            uchar t0 = src0[i];
-            uchar t1 = src1[i];
+            for( ; i < sz.width; i++ )
+            {
+                uchar t0 = srcTop[i];
+                uchar t1 = srcBottom[i];
 
-            dst0[i] = t1;
-            dst1[i] = t0;
+                dstTop[i] = t1;
+                dstBottom[i] = t0;
+            }
         }
     }
+private:
+    const uchar* src0_;
+    const size_t sstep_;
+    uchar* dst0_;
+    const size_t dstep_;
+    const Size size_;
+    const size_t esz_;
+};
+
+static void
+flipVert( const uchar* src0, size_t sstep, uchar* dst0, size_t dstep, Size size, size_t esz )
+{
+    parallel_for_(Range(0, (size.height / 2)), flipVertComputer(src0, sstep, dst0, dstep, size, esz));
```

Comment:
Lets apply this limit for both PRs (similar to [this](https://github.com/opencv/opencv/blob/3.4.15/modules/imgproc/src/color.hpp#L393)):
```
size_t total = esz * size.width * size.height;
double nstripes = (total > (1<<19)) ? (total/(double)(1<<16)) : 1;
parallel_for_(... , nstripes);
```

Reply:
Looking on desktop results, there is no significant speedup besides cases (3840x2160, 8UC1) and (1920x1080, 8UC4). Pi4 version has speedup only for (3840x2160, 8UC1). I'd say that the patch does not bring optimization, but introduce extra overhead (threads) & complexity.

Reply:
Also the speedup is most probably causes by caching details, but not parallel execution.